server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperation.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.purgatory;
+
+import org.apache.kafka.server.util.timer.TimerTask;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * An operation whose processing needs to be delayed for at most the given delayMs. For example
+ * a delayed produce operation could be waiting for specified number of acks; or
+ * a delayed fetch operation could be waiting for a given number of bytes to accumulate.
+ * <br/>
+ * The logic upon completing a delayed operation is defined in onComplete() and will be called exactly once.
+ * Once an operation is completed, isCompleted() will return true. onComplete() can be triggered by either
+ * forceComplete(), which forces calling onComplete() after delayMs if the operation is not yet completed,
+ * or tryComplete(), which first checks if the operation can be completed or not now, and if yes calls
+ * forceComplete().
+ * <br/>
+ * A subclass of DelayedOperation needs to provide an implementation of both onComplete() and tryComplete().
+ * <br/>
+ * Noted that if you add a future delayed operation that calls ReplicaManager.appendRecords() in onComplete()
+ * like DelayedJoin, you must be aware that this operation's onExpiration() needs to call actionQueue.tryCompleteAction().
+ */
+public abstract class DelayedOperation extends TimerTask {
+
+    private final AtomicBoolean completed = new AtomicBoolean(false);
+    // Visible for testing
+    final Lock lock;
+
+    public DelayedOperation(long delayMs, Optional<Lock> lockOpt) {
```

Comment:
It seems that the only caller to this constructor is the one in ReplicaManager. Could you change ReplicaManager to use one of the other constructors depending on lockOpt? Then, we could remove this constructor.

Reply:
It seems like there's a single code path `GroupMetadataManager.appendForGroup()` that actually sets the lock. I've not looked very closely yet but I think we may be able to remove that altogether as the purgatory should not need to reuse the lock from `GroupMetadataManager`. But I'd rather do that refactoring in a separate PR and keep this constructor for now.